src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,1489 @@
+    MakeAlias(layerOutputs[i], layerOutputMatrix, layerOutputSize,
+      batchSize, offset);
+    offset += batchSize * layerOutputSize;
+    sortedIndices.insert({ sortedNetwork[i], i });
+  }
+
+  sortedIndices.insert({
+    sortedNetwork.back(), sortedNetwork.size() - 1
+  });
+
+  //setup layerInputs
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    Layer<MatType>* layer = network[currentLayer];
+    const std::vector<size_t>& parents = parentsList[currentLayer];
+    size_t numParents = parents.size();
+
+    if (numParents == 1)
+    {
+      size_t parentIndex = sortedIndices[parents.front()];
+      MakeAlias(layerInputs[i - 1], layerOutputs[parentIndex],
+        layerOutputs[parentIndex].n_rows, layerOutputs[parentIndex].n_cols, 0);
+    }
+    else
+    {
+      size_t concatSize = layer->InputDimensions()[0];
+      for (size_t j = 1; j < layer->InputDimensions().size(); j++)
+        concatSize *= layer->InputDimensions()[j];
+      MakeAlias(layerInputs[i - 1], layerOutputMatrix, concatSize,
+        batchSize, offset);
+      offset += concatSize;
+    }
+  }
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::InitializeBackwardPassMemory(const size_t batchSize)
+{
+  CheckNetwork("DAGNetwork::InitializeBackwardPassMemory()", 0);
+
+  if (network.size() <= 1)
+    return;
+
+  size_t deltaMatrixSize = 0;
+  for (size_t i = 0; i < sortedNetwork.size(); i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    Layer<MatType>* layer = network[currentLayer];
+    size_t layerDeltaSize = layer->OutputSize();
+    size_t concatDeltaSize = 0;
+
+    if (childrenList[currentLayer].size() > 1)
+      layerDeltaSize *= 2;
+
+    if (parentsList[currentLayer].size() > 1)
+    {
+      concatDeltaSize = layer->InputDimensions()[0];
+      for (size_t j = 1; j < layer->InputDimensions().size(); j++)
+        concatDeltaSize *= layer->InputDimensions()[j];
+    }
+
+    // If at the last layer, we only need to add it's `concatDeltaSize`.
+    if (i == sortedNetwork.size() - 1)
+      layerDeltaSize = 0;
+
+    deltaMatrixSize += layerDeltaSize + concatDeltaSize;
+  }
+
+  if (batchSize * deltaMatrixSize == layerDeltaMatrix.n_elem)
+    return;
+
+  DeleteExtraDeltas();
+  layerDeltaMatrix = MatType(1, batchSize * deltaMatrixSize);
+  outputDeltas.clear();
+  accumulatedDeltas.clear();
+  inputDeltas.clear();
+
+  size_t offset = 0;
+  for (size_t i = 0; i < sortedNetwork.size() - 1; i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+
+    if (childrenList[currentLayer].size() > 1)
+    {
+      accumulatedDeltas.insert({ i, &layerDeltas[i] });
+      MakeAlias(*accumulatedDeltas[i], layerDeltaMatrix,
+        network[currentLayer]->OutputSize(), batchSize, offset);
+
+      outputDeltas.insert({ i, new MatType() });
```

Comment:
Does `outputDeltas` really need to hold pointers to matrices?  Since all the matrices are aliases anyway, couldn't we just have `outputDeltas` hold a `MatType` internally and then set the alias correctly?  If so, then when we did `outputDeltas.clear()` all of the memory issues would be handled.  (The same for the other structures.)

Reply:
If you do `insert({ i, std::move(layerDeltas[i]) })` I think that might preserve the alias?  Or it could be restructured such that you create the aliases after all the `outputDeltas.insert()` calls?